File: app/src/main/java/com/phoneguard/blocker/GuardPrefs.java
package com.phoneguard.blocker;

import android.content.Context;
import android.content.SharedPreferences;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.TimeUnit;

public class GuardPrefs {
    private static final String PREF = "phoneguard_prefs";
    private static final String KEY_KEYWORDS = "keywords";
    private static final String KEY_BLOCKED_ACTIVE = "blocked_active";
    private static final String KEY_STREAK_START = "streak_start_millis";

    public static Set<String> getKeywords(Context ctx) {
        SharedPreferences sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE);
        Set<String> saved = sp.getStringSet(KEY_KEYWORDS, null);
        
        if (saved == null || saved.isEmpty()) {
            Set<String> defaults = new HashSet<>();
            defaults.add("porn");
            defaults.add("xvideos");
            defaults.add("xnxx");
            defaults.add("xhamster");
            defaults.add("pornhub");
            defaults.add("redtube");
            defaults.add("youporn");
            defaults.add("adult");
            defaults.add("sex");
            defaults.add("nude");
            defaults.add("18+");
            return defaults;
        }
        return saved;
    }

    public static void saveKeywords(Context ctx, String raw) {
        Set<String> set = new HashSet<>();
        for (String line : raw.split("\n")) {
            String trimmed = line.trim().toLowerCase();
            if (!trimmed.isEmpty()) {
                set.add(trimmed);
            }
        }
        SharedPreferences sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE);
        sp.edit().putStringSet(KEY_KEYWORDS, set).apply();
    }

    public static boolean isBlockActive(Context ctx) {
        SharedPreferences sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE);
        return sp.getBoolean(KEY_BLOCKED_ACTIVE, false);
    }

    public static void setBlockActive(Context ctx, boolean active) {
        SharedPreferences sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE);
        sp.edit().putBoolean(KEY_BLOCKED_ACTIVE, active).apply();
    }

    public static void clearBlockOnBoot(Context ctx) {
        setBlockActive(ctx, false);
    }

    public static long getStreakStart(Context ctx) {
        SharedPreferences sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE);
        long start = sp.getLong(KEY_STREAK_START, 0L);
        if (start == 0L) {
            start = System.currentTimeMillis();
            sp.edit().putLong(KEY_STREAK_START, start).apply();
        }
        return start;
    }

    public static void resetStreak(Context ctx) {
        long now = System.currentTimeMillis();
        SharedPreferences sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE);
        sp.edit().putLong(KEY_STREAK_START, now).apply();
    }

    public static long getStreakDays(Context ctx) {
        long start = getStreakStart(ctx);
        long diff = System.currentTimeMillis() - start;
        return TimeUnit.MILLISECONDS.toDays(diff);
    }

    public static int getGoalDays() {
        return 30;
    }
}

File: app/src/main/java/com/phoneguard/blocker/GuardPrefs.kt
package com.phoneguard.blocker

import android.content.Context
import java.util.concurrent.TimeUnit

object GuardPrefs {
    private const val PREF = "phoneguard_prefs"
    private const val KEY_KEYWORDS = "keywords"
    private const val KEY_BLOCKED_ACTIVE = "blocked_active"          // resets only on reboot
    private const val KEY_STREAK_START = "streak_start_millis"
    private const val KEY_LAST_RELAPSE = "last_relapse_millis"

    private val DEFAULT_KEYWORDS = setOf(
        "porn", "xvideos", "xnxx", "xhamster", "pornhub", "redtube",
        "youporn", "adult video", "sex video", "nude", "18+", "hentai"
    )

    fun getKeywords(ctx: Context): Set<String> {
        val sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE)
        val saved = sp.getStringSet(KEY_KEYWORDS, null)
        return if (saved.isNullOrEmpty()) DEFAULT_KEYWORDS else saved
    }

    fun saveKeywords(ctx: Context, raw: String) {
        val set = raw.split("\n")
            .map { it.trim().lowercase() }
            .filter { it.isNotEmpty() }
            .toSet()
        ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE)
            .edit().putStringSet(KEY_KEYWORDS, set).apply()
    }

    // ----- Block state: once triggered, stays "active" (locked) until device reboot -----
    fun isBlockActive(ctx: Context): Boolean =
        ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE).getBoolean(KEY_BLOCKED_ACTIVE, false)

    fun setBlockActive(ctx: Context, active: Boolean) {
        ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE)
            .edit().putBoolean(KEY_BLOCKED_ACTIVE, active).apply()
    }

    // Called by BootReceiver — clears the "lock" so cat screen can appear fresh next time,
    // but ALSO counts as a relapse-reset point for the streak per user's spec
    // (kitty only goes away on restart -> restart = the moment user "broke and restarted").
    fun clearBlockOnBoot(ctx: Context) {
        setBlockActive(ctx, false)
    }

    // ----- Streak -----
    fun getStreakStart(ctx: Context): Long {
        val sp = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE)
        var start = sp.getLong(KEY_STREAK_START, 0L)
        if (start == 0L) {
            start = System.currentTimeMillis()
            sp.edit().putLong(KEY_STREAK_START, start).apply()
        }
        return start
    }

    fun resetStreak(ctx: Context) {
        val now = System.currentTimeMillis()
        ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE)
            .edit()
            .putLong(KEY_STREAK_START, now)
            .putLong(KEY_LAST_RELAPSE, now)
            .apply()
    }

    fun getStreakDays(ctx: Context): Long {
        val start = getStreakStart(ctx)
        val diff = System.currentTimeMillis() - start
        return TimeUnit.MILLISECONDS.toDays(diff)
    }

    fun getGoalDays(): Int = 30
}
